phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRebuildRegionScanner.java

```diff
@@ -504,6 +539,75 @@ public Boolean call() throws Exception {
+            keys.add(PVarbinary.INSTANCE.getKeyRange(entry.getKey()));
+            perTaskDataKeyToDataPutMap.put(entry.getValue().getRow(), entry.getValue());
+            if (keys.size() == rowCountPerTask) {
+                addVerifyTask(keys, perTaskDataKeyToDataPutMap);
+                keys = new ArrayList<>(rowCountPerTask);
+                perTaskDataKeyToDataPutMap = Maps.newTreeMap(Bytes.BYTES_COMPARATOR);
+            }
+        }
+        if (keys.size() > 0) {
+            addVerifyTask(keys, perTaskDataKeyToDataPutMap);
+        }
+        List<Boolean> taskResultList = null;
+        try {
+            LOGGER.debug("Waiting on index verify tasks to complete...");
+            taskResultList = this.pool.submitUninterruptible(tasks);
+        } catch (ExecutionException e) {
+            throw new RuntimeException("Should not fail on the results while using a WaitForCompletionTaskRunner", e);
+        } catch (EarlyExitFailure e) {
+            throw new RuntimeException("Stopped while waiting for batch, quitting!", e);
+        } finally {
+            tasks.getTasks().clear();
+        }
+        for (Boolean result : taskResultList) {
+            if (result == null) {
+                // there was a failure
+                throw new IOException(exceptionMessage);
+            }
+        }
+    }
+
+    private void verifyAndOrRebuildIndex() throws IOException {
+        if (verifyType == IndexTool.IndexVerifyType.AFTER || verifyType == IndexTool.IndexVerifyType.NONE ||
+                verifyType == IndexTool.IndexVerifyType.NONE) {
+            // For these options we start with rebuilding index rows
+            rebuildIndexRows(mutations);
+        }
+        if (verifyType == IndexTool.IndexVerifyType.NONE) {
+            return;
+        }
+        if (verifyType == IndexTool.IndexVerifyType.BEFORE || verifyType == IndexTool.IndexVerifyType.BOTH ||
+                verifyType == IndexTool.IndexVerifyType.ONLY) {
+            // For these options we start with verifying index rows
+            onlyVerify = true; // Don't stop at the first mismatch
+            parallelizeIndexVerify();
+        }
+        if (verifyType == IndexTool.IndexVerifyType.BEFORE || verifyType == IndexTool.IndexVerifyType.BOTH) {
+            // For these options, we have identified the rows to be rebuilt and now need to rebuild them
+            // At this point, dataKeyToDataPutMap includes mapping only for the rows to be rebuilt
+            mutations.clear();
+            for (Map.Entry<byte[], Put> entry: dataKeyToDataPutMap.entrySet()) {
+                mutations.add(entry.getValue());
+            }
+            rebuildIndexRows(mutations);
+        }
+
+        if (verifyType == IndexTool.IndexVerifyType.AFTER || verifyType == IndexTool.IndexVerifyType.BOTH) {
+            // We have rebuilt index row and now we need to verify them
+            onlyVerify = false; // Stop at the first mismatch
```

Comment:
if onlyVerify now really means "don't stop at first mismatch" whether we're doing rebuilds or not, then the variable name is confusing.

Reply:
I rename it to doNotFail